Test/QuickCheck/Function.hs

```diff
@@ -452,20 +456,65 @@ shrinkFun shr (Map g h p) =
 data Fun a b = Fun (a :-> b, b, Bool) (a -> b)
 
 #if defined(__GLASGOW_HASKELL__) && __GLASGOW_HASKELL__ >= 708
--- | A pattern for matching against the function only:
+-- pattern Fn :: (a -> b) -> Fun a b
+
+-- | A modifier for testing functions.
 --
 -- > prop :: Fun String Integer -> Bool
 -- > prop (Fn f) = f "banana" == f "monkey"
 -- >            || f "banana" == f "elephant"
 pattern Fn f <- Fun _ f
+
+-- pattern Fn2 :: (a -> b -> c) -> Fun (a, b) c
+
+-- | A modifier for testing binary functions.
+-- 
+-- @ 
+--     prop_zipWith :: Fun (Int, Bool) Char -> [Int] -> [Bool] -> Bool
+--     prop_zipWith (Fn2 f) xs ys = zipWith f xs ys == [ f x y | (x, y) <- zip xs ys]
+-- @
+--
+-- >>> quickCheck prop_zipWith
+-- +++ OK, passed 100 tests.
+-- 
+pattern Fn2 :: (a -> b -> c) -> Fun (a, b) c
```

Comment:
Type signatures for patterns works only with `GHC 7.10+` (it's omitted for `Fn` for that reason)